File: go/leetcode/647-palindromic-substrings.go
package main

func countSubstrings(s string) int {
	n := len(s)

	if n < 2 {
		return n
	}

	count := 0

	expand := func(l, r int) {
		for l >= 0 && r < n && s[l] == s[r] {
			count++
			l--
			r++
		}
	}

	for i := range s {
		expand(i, i)
		expand(i, i+1)
	}

	return count
}

File: ts/leetcode/647-palindromic-substrings.ts
function countSubstrings(s: string): number {
    const n = s.length;

    if (n < 2) return n;

    let count = 0;

    const expandAroundCenter = (left: number, right: number) => {
        while (left >= 0 && right < n && s[left] === s[right]) {
            count++;
            left--;
            right++;
        }
    };

    for (let i = 0; i < n; i++) {
        // Odd length palindromes
        expandAroundCenter(i, i);
        // Even length palindromes
        expandAroundCenter(i, i + 1);
    }

    return count;
}